### DISCUSSION TITLE
Pageable 사용 및 페이지 인덱스 시작범위에 대해

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#173] [뜨거운 토론]
> 다음 의견으로 결론이 난 사안입니다.
> - https://github.com/woowacourse-teams/2023-pium/discussions/173#discussioncomment-6611334

## 논제

> 아래 PR로부터 발생한 논제입니다.
- #149 

페이징 과정에서 클라이언트의 요청과 서버에서 반환되는 값이 차이가 나는 현상이 발생했습니다.

### 이유 파악하기

JPARepository에서는 Pageable 객체를 이용하여 페이징된 데이터를 손쉽게 가공할 수 있도록 제공하고있습니다.

사용 방법은 다음과 같습니다.

```java
public interface HistoryRepository extends JpaRepository<History, Long> {

    Page<History> findAllByPetPlantId(Long petPlantId, Pageable pageable);
}
```

이 때 JPA에서는 페이징을 조회할 때 사용하는 기본 전략이 `index를 0부터 센다`입니다.

때문에 클라이언트와 1-based 인덱싱으로 진행하도록 협의를 본 현 상황에서 요청이 들어오고 반환될 때 각각 값을 정제해줘야합니다.

### 현재 문제상황

1. `spring.data.web.pageable.one-indexed-parameters=true`에 대한 오해

`spring.data.web.pageable.one-indexed-parameters=true` 설정은 HTTP 통신으로부터 생성되는 Pageable 객체에 대해 1-based indexing을 적용합니다.
때문에 HTTP 통신을 거치치 않는 서비스 레이어에서의 통합테스트는 위 설정이 적용되지 않습니다.

2. 파라미터로 Pageable 객체를 받고있지 않다.

위 이야기와 이어서 파라미터로 Pageable 객체를 받고있지 않기 때문에 `spring.data.web.pageable.one-indexed-parameters=true`설정이 적용되지 않습니다.

3. Request는 그대로 Response는 +1

위와 같은 상황들이 얽혀서 Reqeust를 받을 때는 값을 정제하지 않고있으나 Response 시에만 page + 1을 하는 작업이 적용되어있습니다.

따라서 클라이언트가 /history?page=1 로 요청했을 때 반환되는 page는 2가 되어버립니다.

### 해결책

> 1-based 인덱싱을 적용한다는 것을 기본적인 전제로 설정하고 해결책을 서술합니다.

이에 대한 해결책은 크게 두가지 방법으로 나뉩니다.

### 1. 들어오고 나오는 값에 대해 일일히 정제하는 방법

Requset로 page에 대한 요청이 들어왔을 때 -1을 해줍니다.
-1된 page 정보를 기준으로 Repository에 조회를 요청합니다.

이후 반환된 값에 대해 page +1을 해주고 클라이언트에게 반환합니다.

```text
Request -> page -1 -> 값 조회

값 조회 -> page + 1 -> Response
```

이 경우 `spring.data.web.pageable.one-indexed-parameters=true` 설정은 필요 없습니다.

* 현재 코드 상황으로는 Request에서 `page-1`을 해준다 작업만 추가되면 되는 상황입니다.

### 2. Request에서 Pageable로 값 받기

application.properties에 `spring.data.web.pageable.one-indexed-parameters=true` 설정을 적용합니다.

이후 Controller에서 페이징에 대한 요청을 받을 때 Pageable 객체로 요청을 받습니다.

이 경우 Presentation Layer (Controller단) 에서 Pageable 객체가 생성되면서 1-based 인덱싱으로 관리해줍니다.

- 요청값
![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/c59b7cf5-4f5d-4527-b21b-0d90726d93b0)

- Pageable에 들어왔을 때 0으로 받는다
![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/665b6139-a6f3-4c95-bb10-2f3e09a82da5)

**Comment by Kim0914** [comment]
저는 우선 `spring.data.web.pageable.one-indexed-parameters`를 삭제하고 서비스 계층에서 인덱싱을 해주는게 좋지 않을까 싶습니다.

`spring.data.web.pageable.one-indexed-parameters` 속성이 Pagable 객체로 바인딩되는 과정에만 영향을 끼치고, response에는 결국 +1을 해줘야 하기 때문이죠.

Pagable을 사용하든 현재처럼 직접 PageRequest를 만들어주던 구애받지 않고 service에서 -1을 해주면 해결될 것 같네요 !
응답도 공통적으로 +1을 해줘야 하구요.

그럼, 마지막으로 논의해볼 사항이 Pagable을 사용할 것이냐, 지금처럼 유지할 것이냐를 이야기 해볼 수 있겠네요.

**Comment by Choi-JJunho** [comment]
저는 개인적으로 `2. Request에서 Pageable로 값 받기` 방식을 선호합니다.

이유는 다음과 같습니다.

## 1. 들어오고 나오는 값에 대해 일일히 정제하는 방법

![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/a4726bbc-8fff-444d-8869-3b2de31b1d68)

### 장점
- page 값을 직관적으로 알아보기 쉽다
- 컨트롤러에서 각각의 인자에 대해 명시적인 Validation을 수행할 수 있다.

### 단점
- 코드의 재사용성이 떨어진다
- 페이징을 하는 코드가 많아질수록 중복 코드가 많이 생길 우려가 있다
- page 요청이 없을 때 기본값을 지정하기가 번거롭다

`page 요청이 없을 때 기본값을 지정할 수 없다`에 대한 추가적인 의견 덧붙이겠습니다.
개인적으로 쿼리 파라미터는 **질의의 영역**이기 때문에 클라이언트에게 필수적으로 요구해야하는 값은 아니라고 생각합니다.
이 때문에 `/history` 와 같은 요청이 왔을 때 기본값을 지정해줘야한다고 생각합니다.

현재 구조 상으로는 강제로 요구하고있기 때문에 위와같은 의견과 상충된다고 생각합니다

## 2. Request에서 Pageable로 값 받기

![image](https://github.com/woowacourse-teams/2023-pium/assets/49794401/c62b7a5a-6d12-4d5b-80fb-d4e8ecf79a79)

### 장점
- Pageable 객체를 인자로 받으면서 파라미터에서 관리할 값이 줄어든다.
  -  page, size 값을 인자로 일일히 받아오던 부분을 객체 하나로 관리할 수 있습니다.
- 향후 정렬조건에 대한 부분을 클라이언트에서 제어할 수 있다.
  - 시간순, 이름순, 마지막으로 물 준 날짜 순 / 오름차순, 내림차순
- default 값을 Pageable 수준에서 정의할 수 있다.
- 코드의 재사용성 및 중복이 1번에 비해 적다
  - 향후 페이징을 필요로하는 기능이 존재할 때 Controller에 Pageable 객체를 사용한다는 부분외에는 중복되는 코드가 없다는 것을 기대할 수 있습니다.

### 단점
- 각 속성(page, size) 에 대한 검증을 설정하는 것이 1번 방법보다는 다소 번거로울 수 있다. (ArgumentResolver를 이용하여 가능합니다.)
- `spring.data.web.pageable.one-indexed-parameters=true` 속성을 사용할 경우 다소 헷갈릴 수 있다.

저는 팀원들의 싱크가 맞은 상태에서 `spring.data.web.pageable.one-indexed-parameters=true` 속성을 사용하면서 Response에만 +1한 값을 넘겨줘도 괜찮다고 생각합니다.

하지만 추후 코드의 유지보수성을 생각해봤을 때는 위에서 그레이가 말한 것 처럼  `spring.data.web.pageable.one-indexed-parameters=true` 속성을 제거하고 값을 정제하는 방향도 좋다고 생각합니다.


+) 페이지, 사이즈에 대한 검증이 필요할까? 에 대한 의문도 들지만 이 내용은 본 discussion에 대한 논의가 마무리 된 후에 이야기하는것이 좋아보입니다.

**Comment by rawfishthelgh** [comment]
일단 `one-indexed-parameters`설정을 제거하자는 의견에 동의합니다.
"페이지 인덱싱을 1로부터 시작한다" 라는 관심사가 프로덕션 코드에 있지 않고 프로퍼티 설정에 있는 것이 어색하다는 생각이 드네요!

또 Pageable 객체를 통해 값을 바인딩하는 것이 코드 재사용성과 유연하게 질의를 하는 측면에서 더 적합해보입니다. 
ArgumentResolver를 통해 이를 검증하는 법이 있다고 하니 이를 통해 해결하면 될 것 같습니다. 
그렇다면 0번째부터 인덱스가 시작하게 될 텐데, 두 가지가 고민이 되네요.

1. 컨트롤러 단 혹은 서비스 레이어에서 인덱싱을 -1 해주는 작업, 또 response에서 인덱싱을 +1 해주는 작업을 수행한다
2. 인덱스를 0부터 시작하는 것으로 변경한다

저는 2번 방법이 조금 더 적합하다고 생각도 듭니다
사실 저희가 구현한 "무한 스크롤"이 클라이언트 입장에서 어떠한 "특정 페이지 번호"를 선택해 이동하는 것이 아니고,
단순히 스크롤을 죽 내려서 순차적으로 값을 불러오는 상황입니다.
5번 페이지에서 다시 2번으로 갔다가 오는 상황이 일어나지 않습니다.
(그냥 0, 1, 2, 3, 4 순서대로 갈 수 밖에 없음)

그렇다면 쿼리 파라미터 인덱싱을 몇 부터 시작할지를 지정하는게 큰 의미가 없고, 가장 부수효과가 적은 방법(0부터 시작)을 선택하는게 안전하다고 봅니다.

**Comment by yeonkkk** [comment]
만약 페이지 인덱싱을 1부터 시작한다면 저는 1번 방법을 가져가는 것도 괜찮은 것 같아요. 간단하고 보다 직관적일 것 같네요.

그리고 두 가지 방법을 놓고 봤을 때, 1번 방법이 기본값 지정이 어렵고 상대적으로 중복 코드가 많이 발생할 수 있지만 현재 기획 내에서는 큰 문제가 되지 않는다고 생각됩니다. 
히스토리 조회 말고도 리마인더 조회, 반려식물 전체 조회 등 페이징을 적용할 수 있는 기능들이 많지만 적용하지 않기로 했으니까요.
request에서의 pageable 바인딩은 이후에 기획이 변경되어 페이징이 많이 필요해져서 중복 코드가 증가한다면 그 때 변경해도 좋을 것 같습니다.

그리고 하마드의 의견에 동의합니다. 
저희가 구현하는 기능이 무한 스크롤이기 때문에 인덱싱을 0부터 시작해도 큰 문제가 없다고 생각합니다.
1로 시작하면 양측이 어떤 방법을 사용하든 +1해주는 작업이 필요하기 때문에 꼭 필요하지 않다면 0으로 하는 것이 더 좋을 것 같습니다.
이 부분은 이전에 페이징에 대한 회의를 진행했을 때, 프론트엔드에서도 이야기했던 부분이라서 고려해볼만한 문제인 것 같습니다.

**Comment by rawfishthelgh** [comment]
# 결론
1. 페이징 인덱스 시작점을 0부터 시작하는 것으로 한다
2. Pageable 객체로 바인딩하는 것으로 한다
3. 디폴트값을 (page = 0, size = 10, "waterDate", DESC)지정하고, 쿼리 파라미터가 없을 때 응답한다. 검증은 수행하지 않는다.

**Comment by Choi-JJunho** [comment]
> # 결론
> 1. 페이징 인덱스 시작점을 0부터 시작하는 것으로 한다
> 2. Pageable 객체로 바인딩하는 것으로 한다
> 3. 디폴트값을 (page = 0, size = 10, "waterDate", DESC)지정하고, 쿼리 파라미터가 없을 때 응답한다. 검증은 수행하지 않는다.

해당 내용 적용 완료하여 close하겠습니다
- https://github.com/woowacourse-teams/2023-pium/pull/149
